drv/lpc55-sprot-server/src/main.rs

```diff
@@ -175,280 +114,618 @@ fn main() -> ! {
         }
+    }
+}
 
-        // Prime FIFOWR in order to be ready for start of frame.
-        //
-        // All our interrupts are left enabled for the sake of simplicity.
-        // The downside is that the following drain will elicit a spurious
-        // interrupt. But, that interrupt will occur before the start of any
-        // frame when we are not in a realtime situation.
-        self.spi.drain_tx(); // FIFOWR is now empty; we'll get an interrupt.
-        loop {
-            if self.txcount >= tx_end || !self.spi.can_tx() {
-                break;
+impl Io {
+    /// Put a Protocol::Busy value in FIFOWR so that SP/logic analyzer knows
+    /// we're away.
+    ///
+    /// This is primarily for debugging via logic analyzer, as it's only done
+    /// after reading a request but before sending a response. The SP won't
+    /// process a response unless an ROT_IRQ is raised, so it will never notice
+    /// this byte. That is actually what we want as the rest of the handling is
+    /// done correctly on the RoT side. If the SP actually reacted to this byte
+    /// then it would alter its behavior to deal with it, but the only way it
+    /// would see it is if it already clocked a CSn assert. Once that occurs,
+    /// the RoT will notice an overrun and report it to the SP anyway. No need
+    /// to pulse or do anything special here.
+    ///
+    /// At the same time, it's nice to see this byte on the logic analyzer to
+    /// indicate what's going on.
+    fn mark_busy(&mut self) {
+        self.spi.drain_tx();
+        self.spi.send_u8(Protocol::Busy as u8);
+    }
+
+    /// Write a response to the SP
+    ///
+    /// We expect to only clock in 0s from the SP, or else the SP and RoT are
+    /// desynchronized.
+    pub fn spi_write<'a>(&mut self, response: VerifiedTxMsg) {
+        self.state = State::Write(WriteState::WaitingToTransmit);
+
+        // Prime the fifo with the first part of the response to prevent
+        // underrun while waiting for an interrupt.
+        self.spi.drain_tx();
+        let mut tx_iter = response.iter();
+        while self.spi.can_tx() {
+            if let Some(b) = tx_iter.next() {
+                self.spi.send_u8(b);
+            } else {
+                self.spi.send_u8(0);
             }
-            let b = self.tx_buf.as_slice()[self.txcount];
-            self.spi.send_u8(b);
-            self.txcount += 1;
         }
-
-        if transmit {
-            ringbuf_entry!(Trace::Transmit(self.txcount, tx_end));
-            ringbuf_entry!(Trace::RotIrqAssert);
-            self.gpio.set_val(ROT_IRQ, Value::Zero).unwrap_lite();
+        let fifostat = self.spi.fifostat();
+        ringbuf_entry!(Trace::FifoStat(
+            fifostat.txlvl().bits(),
+            fifostat.rxlvl().bits(),
+            fifostat.txerr().bits(),
+            fifostat.rxerr().bits()
+        ));
+
+        self.assert_rot_irq();
+        loop {
+            sys_irq_control(SPI_IRQ, true);
```

Comment:
Ah, that makes sense. Thanks.

This looks like a problem on master as well: https://github.com/oxidecomputer/hubris/blob/master/drv/lpc55-sprot-server/src/main.rs#L296-L316

Reply:
I'm not so sure about this. The reason the code is written this way here and on master is that the ROT_IRQ is asserted outside the loop, once. Otherwise we'd be asserting it every time through the loop. Is that what we want to do?

Reply:
I put the ROT_IRQ inside the loop as suggested. I'm still not sure this is the correct choice.